### PR TITLE
Reuse document meta store state from prepare step

### PR DESCRIPTION
**Pull request by toregge**
@baldersheim : please review

### CODE REVIEW

File: searchcore/src/vespa/searchcore/proton/documentmetastore/i_store.h [hunk 1]
```diff
@@ -6,6 +6,7 @@
 #include <vespa/document/base/globalid.h>
 #include <vespa/document/bucket/bucketid.h>
 #include <persistence/spi/types.h>
+#include <vespa/searchlib/common/serialnum.h>
```
Comment (baldersheim) [inline searchcore/src/vespa/searchcore/proton/documentmetastore/i_store.h:9]: Do you need this when you are using uint64_t instead of SerialNum

Comment (toregge) [reply searchcore/src/vespa/searchcore/proton/documentmetastore/i_store.h]: No. Removed.

File: searchcore/src/vespa/searchcore/proton/documentmetastore/documentmetastore.cpp [hunk 2]
```diff
@@ -536,40 +538,44 @@ DocumentMetaStore::updateMetaData(DocId lid,
...
                         oldMetaData.getTimestamp(), oldMetaData.getDocSize(),
                         _subDbType);
-    return true;
 }
 
 bool
-DocumentMetaStore::remove(DocId lid)
+DocumentMetaStore::remove(DocId lid, uint64_t prepare_serial_num)
 {
     BucketDBOwner::Guard bucketGuard = _bucketDB->takeGuard();
```
Comment (baldersheim) [inline searchcore/src/vespa/searchcore/proton/documentmetastore/documentmetastore.cpp:569]: Wait with takeGuard untial after validLid check ?

Comment (toregge) [reply searchcore/src/vespa/searchcore/proton/documentmetastore/documentmetastore.cpp]: Fixed.

File: searchcore/src/vespa/searchcore/proton/documentmetastore/documentmetastore.cpp [hunk 3]
```diff
@@ -462,20 +462,23 @@ DocumentMetaStore::put(const GlobalId &gid,
...
+                       DocId lid,
+                       uint64_t prepare_serial_num)
 {
     Result res;
     RawDocumentMetaData metaData(gid, bucketId, timestamp, docSize);
     KeyComp comp(metaData, _metaDataStore, *_gidCompare);
-    TreeType::Iterator itr = _gidToLidMap.lowerBound(KeyComp::FIND_DOC_ID,
-            comp);
+    auto& itr = _gid_to_lid_map_write_itr;
+    if (prepare_serial_num == 0u || _gid_to_lid_map_write_itr_prepare_serial_num != prepare_serial_num) {
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/documentmetastore/documentmetastore.cpp:472]: Consider adding a helper function for this, e.g. update_gid_to_lid_map_write_itr(prepare_serial_num). Same check is used in three places.

File: searchcore/src/vespa/searchcore/proton/documentmetastore/documentmetastore.h [hunk 4]
```diff
@@ -68,6 +68,8 @@ class DocumentMetaStore final : public DocumentMetaStoreAttribute,
 
     MetaDataStore       _metaDataStore;
     TreeType            _gidToLidMap;
+    Iterator            _gid_to_lid_map_write_itr; // Iterator used for all updates of _gidToLidMap
+    SerialNum           _gid_to_lid_map_write_itr_prepare_serial_num;
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/documentmetastore/documentmetastore.h:72]: Consider adding a comment on how this serial num relates to _gid_to_lid_map_write_itr.

**Review by geirst — COMMENTED** [review]
👍  Some comments to consider

**Pull request merged**